csharp/src/Google.Protobuf/Collections/MapField.cs

```diff
@@ -221,17 +227,15 @@ public bool TryGetValue(TKey key, out TValue value)
             }
         }
 
-        // TODO: Make these views?
-
         /// <summary>
         /// Gets a collection containing the keys in the map.
         /// </summary>
-        public ICollection<TKey> Keys { get { return list.Select(t => t.Key).ToList(); } }
+        public ICollection<TKey> Keys { get { return new MapView<TKey>(this, pair => pair.Key, ContainsKey); } }
```

Comment:
Okay - I don't much mind either way, although I don't have that expectation
and of course it's only detectable via reference equality. (If the returned
objects had some other difference, I would be much more in favour.)

Will do that at the same time as the CopyTo fix.

Jon


On 10 August 2015 at 07:10, Jan Tattermusch <notifications@github.com>
wrote:

> In csharp/src/Google.Protobuf/Collections/MapField.cs
> <https://github.com/google/protobuf/pull/701#discussion_r36603997>:
>
> >          /// <summary>
> >          /// Gets a collection containing the keys in the map.
> >          /// </summary>
> > -        public ICollection<TKey> Keys { get { return list.Select(t => t.Key).ToList(); } }
> > +        public ICollection<TKey> Keys { get { return new MapView<TKey>(this, pair => pair.Key, ContainsKey); } }
>
> The benefit -
> I think always obtaining the same object once you access property like
> Keys or Values (or property in general) is just what everyone would expect.
> And breaking natural expectations is almost always a bad idea. Let's favor
> API that behaves in the expected way over saving 16 bytes.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/google/protobuf/pull/701/files#r36603997>.
>